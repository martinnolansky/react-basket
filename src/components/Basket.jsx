import React from "react";

const Basket = ({ items, removeFromBasket }) => {
  // Run removeFromBasket in App.js if remove button clicked
  const handleRemoveClick = (productKey) => {
    removeFromBasket(productKey);
  };

  // Calculate total amount of items in basket
  const totalAmount = items.reduce(
    (total, item) => total + parseFloat(item.monthlyCost || 0),
    0
  );

  // Render the basket items, updating new mapped array when items are added or removed from
  return (
    <aside aria-live="polite">
      <div className="border-8 border-black inline-block p-4">
        <h2 className="mb-4" aria-level="2">
          Basket
        </h2>
        <ul>
          {items.map((item, index) => (
            <li key={index} data-testid={`basket-item-${item.keyProp}`}>
              {item.productName} - £{item.monthlyCost} per month
              <button
                onClick={() => handleRemoveClick(item.keyProp)}
                data-testid={`remove-button-${item.keyProp}`}
                className="ml-2 text-red-500"
              >
                Remove
              </button>
            </li>
          ))}
        </ul>
        {items.length === 0 && (
          <p data-test-id="empty-basket-message" role="status">
            Your basket is empty.
          </p>
        )}
        <p className="mt-4 font-bold" role="status">
          Total Amount: £{totalAmount.toFixed(2)}
        </p>
      </div>
    </aside>
  );
};

export default Basket;
