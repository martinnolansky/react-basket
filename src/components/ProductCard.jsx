import React from "react";
import PropTypes from "prop-types";

// Destructure the ProductCard component with props that are passed from App.js
const ProductCard = (props) => {
  const {
    productName,
    src,
    description,
    monthlyCost,
    contractLength,
    keyProp,
    type,
    addToBasket,
    removeFromBasket,
    isInBasket,
  } = props;

  // On button click, run removeFromBasket if item is in basket or addToBasket if item is not in basket, passing in unique keyProp ID
  const buttonClick = () => {
    if (isInBasket) {
      removeFromBasket(keyProp);
    } else {
      addToBasket({
        productName,
        monthlyCost,
        contractLength,
        type,
        keyProp,
      });
    }
  };

  // Return the JSX for each productCard, displaying the product details and updating the button dynamically depending if the product is in the basket
  return (
    <>
      <div className="max-w-sm bg-white border border-gray-200 rounded-lg shadow dark:bg-gray-800 dark:border-gray-700">
        <img
          className="rounded-t-lg mx-auto"
          src={src}
          alt={`${productName} Logo`}
          style={{ maxWidth: "100%", height: "auto" }}
        />
        <div className="p-5">
          <h5 className="mb-2 text-2xl font-bold tracking-tight text-gray-900 dark:text-white">
            {productName}
          </h5>
          <p className="mb-3 font-normal text-gray-700 dark:text-gray-400">
            {description}
          </p>
          <p className="mb-3 font-normal text-gray-700 dark:text-gray-400">
            Monthly Cost: Â£{monthlyCost}
          </p>
          <p className="mb-3 font-normal text-gray-700 dark:text-gray-400">
            Contract Length: {contractLength} months
          </p>
          <button
            onClick={buttonClick}
            className="inline-flex items-center px-3 py-2 text-sm font-medium text-center text-white bg-blue-700 rounded-lg hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
          >
            {isInBasket ? "Remove from Basket" : "Add to Basket"}
          </button>
        </div>
      </div>
    </>
  );
};

// Define propTypes to ensure type safety of props that are passed to ProductCard Component
ProductCard.propTypes = {
  productName: PropTypes.string.isRequired,
  src: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  monthlyCost: PropTypes.number.isRequired,
  contractLength: PropTypes.number.isRequired,
  keyProp: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  addToBasket: PropTypes.func.isRequired,
  removeFromBasket: PropTypes.func.isRequired,
  isInBasket: PropTypes.bool.isRequired,
};

export default ProductCard;
