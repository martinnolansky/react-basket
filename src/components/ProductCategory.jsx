import React from "react";

// Display the name of each product category with props passed in from App.js
const ProductCategory = (props) => {
  const { name } = props;
  return (
    <header>
      <h2 className="text-2xl font-semibold mt-12" aria-level="2">
        {name}
      </h2>
    </header>
  );
};

export default ProductCategory;
