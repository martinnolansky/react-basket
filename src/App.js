import React, { useState } from "react";
import { ToastContainer, toast } from "react-toastify";
import "react-toastify/dist/ReactToastify.css";
import "./App.css";
import ProductCard from "./components/ProductCard";
import ProductCategory from "./components/ProductCategory";
import Basket from "./components/Basket";
import productData from "./data/products.json";
import skyF1Image from "./assets/images/SkySportsF1.png";
import SkySportsGolf from "./assets/images/SkySportsGolf.png";
import SkySportsTennis from "./assets/images/SkySportsTennis.png";
import SkySportsFootball from "./assets/images/SkySportsFootball.png";
import SkyHorror from "./assets/images/SkyHorror.png";
import SkyCinema from "./assets/images/SkyCinema.png";

function App() {
  // Group products by type
  const productsByType = {};
  productData.forEach((product) => {
    if (!productsByType[product.type]) {
      productsByType[product.type] = [];
    }
    productsByType[product.type].push(product);
  });

  // Initialize basket to empty array
  const [basketItems, setBasketItems] = useState([]);

  // Add product to basket
  const addToBasket = (productInfo) => {
    const productKey = `${productInfo.type}_${productInfo.keyProp}`;
    const existingProductIndex = basketItems.findIndex(
      (item) => item.productKey === productKey
    );
    if (existingProductIndex === -1) {
      setBasketItems([...basketItems, { ...productInfo, productKey }]);
    } else {
      // Show error toast if product is already in the basket
      toast.error(`${productInfo.productName} is already in the basket`, {
        position: "top-center",
        autoClose: 3000,
        hideProgressBar: false,
        closeOnClick: true,
        pauseOnHover: true,
        draggable: true,
      });
    }
  };

  // Remove product from basket
  const removeFromBasket = (productKey) => {
    const updatedBasket = basketItems.filter(
      (item) => item.keyProp !== productKey
    );
    setBasketItems(updatedBasket);
  };

  // Object to associate product image names with image sources
  const productImages = {
    "Sky F1": skyF1Image,
    "Sky Golf": SkySportsGolf,
    "Sky Tennis": SkySportsTennis,
    "Sky Football": SkySportsFootball,
    "Sky Horror": SkyHorror,
    "Sky Cinema": SkyCinema,
  };

  // Render the app from other components and pass props to them
  return (
    <>
      <h1 className="text-center font-bold mt-4">Sky</h1>
      <div className="flex flex-col md:flex-row h-screen ml-4">
        <div className="flex-grow">
          {Object.entries(productsByType).map(([type, products], index) => (
            <React.Fragment key={type}>
              <ProductCategory name={type} />
              <div
                className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-4"
                role="grid"
              >
                {products.map((product, productIndex) => (
                  <ProductCard
                    key={productIndex}
                    productName={product.name}
                    description={product.description}
                    monthlyCost={parseFloat(product.monthlyCost)}
                    contractLength={parseFloat(product.contractLength)}
                    src={productImages[product.name]}
                    type={product.type}
                    keyProp={`${product.type}_${productIndex}`}
                    addToBasket={addToBasket}
                    removeFromBasket={removeFromBasket}
                    isInBasket={basketItems.some(
                      (item) =>
                        item.keyProp === `${product.type}_${productIndex}`
                    )}
                    role="gridcell"
                  />
                ))}
              </div>
            </React.Fragment>
          ))}
        </div>

        <div className="md:h-screen md:w-1/3 p-4 mt-16" role="complementary">
          <Basket items={basketItems} removeFromBasket={removeFromBasket} />
        </div>
      </div>

      <ToastContainer aria-live="polite" role="status" />
    </>
  );
}

export default App;
