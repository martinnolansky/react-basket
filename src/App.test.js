// Imports
import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import ProductCard from "./components/ProductCard.jsx";
import Basket from "./components/Basket.jsx";
import ProductCategory from "./components/ProductCategory.jsx";
import App from "./App";
import skyF1Image from "./assets/images/SkySportsF1.png";

// Mock functions for addToBasket and removeFromBasket
const mockAddToBasket = jest.fn();
const mockRemoveFromBasket = jest.fn();

// Mock product for tests
const sampleProduct = {
  productName: "Sample Product",
  description: "This is a sample product.",
  monthlyCost: 19.99,
  contractLength: 12,
  keyProp: "sample_key",
  type: "Sample Type",
  src: skyF1Image,
};

// Mock multiple products in basket for tests
const sampleProducts = [
  {
    productName: "Sample Product 2",
    description: "This is a sample product 2.",
    monthlyCost: 19.99,
    contractLength: 12,
    keyProp: "sample_key_2",
    type: "Sample Type",
    src: skyF1Image,
  },
  {
    productName: "Sample Product 3",
    description: "This is a sample product 3.",
    monthlyCost: 19.99,
    contractLength: 12,
    keyProp: "sample_key_3",
    type: "Sample Type",
    src: skyF1Image,
  },
];

/*
This test checks if the ProductCard renders correctly with the provided sample product data.
It verifies that various elements such as product name, description, monthly cost, contract length, and the "Add to Basket" 
button are present in the rendered component. This helps ensure that the component displays the expected content.
calls addToBasket when Add to Basket button is clicked:
*/
test("renders ProductCard correctly", () => {
  render(
    <ProductCard
      {...sampleProduct}
      addToBasket={mockAddToBasket}
      removeFromBasket={mockRemoveFromBasket}
      isInBasket={false}
    />
  );
  // Check if the component renders correctly
  expect(screen.getByText("Sample Product")).toBeInTheDocument();
  expect(screen.getByText("This is a sample product.")).toBeInTheDocument();
  expect(screen.getByText("Monthly Cost: £19.99")).toBeInTheDocument();
  expect(screen.getByText("Contract Length: 12 months")).toBeInTheDocument();
  expect(screen.getByRole("button")).toBeInTheDocument();
});

/*
This test simulates a click on the "Add to Basket" button and checks if the addToBasket function was called with the correct arguments.
It uses fireEvent.click to simulate a button click and then asserts that the mockAddToBasket function was called with the expected product information.
This is a crucial test to verify that the interaction with the "Add to Basket" button works as intended.
*/
test("calls addToBasket when Add to Basket button is clicked", () => {
  render(
    <ProductCard
      {...sampleProduct}
      addToBasket={mockAddToBasket}
      removeFromBasket={mockRemoveFromBasket}
      isInBasket={false}
    />
  );

  // Click the "Add to Basket" button
  fireEvent.click(screen.getByText("Add to Basket"));

  // Check if addToBasket was called with the correct arguments
  expect(mockAddToBasket).toHaveBeenCalledWith({
    productName: "Sample Product",
    monthlyCost: 19.99,
    contractLength: 12,
    type: "Sample Type",
    keyProp: "sample_key",
  });
});

/*
This test simulates a click on the "Remove from Basket" button and checks if the removeFromBasket function was called with the correct arguments.
It uses fireEvent.click to simulate a button click and then asserts that the mockRemoveFromBasket function was called with the expected product information.
This is a crucial test to verify that the interaction with the "Remove from Basket" button works as intended.
*/
test("calls removeFromBasket when Remove from Basket button is clicked", () => {
  render(
    <ProductCard
      {...sampleProduct}
      addToBasket={mockAddToBasket}
      removeFromBasket={mockRemoveFromBasket}
      isInBasket={true} // Set isInBasket to true initially
    />
  );

  // Ensure the "Remove from Basket" button is present
  expect(screen.getByText("Remove from Basket")).toBeInTheDocument();

  // Click the "Remove from Basket" button
  fireEvent.click(screen.getByText("Remove from Basket"));

  // Check if removeFromBasket was called with the correct keyProp
  expect(mockRemoveFromBasket).toHaveBeenCalledWith("sample_key");
});

/*
This test simulates a click on the "Remove" button and checks if the removeFromBasket function was called with the correct arguments.
It uses fireEvent.click to simulate a button click and then asserts that the mockRemoveFromBasket function was called with the expected product information.
This is a crucial test to verify that the interaction with the "Remove" button works as intended.
*/
test("calls removeFromBasket when remove button is clicked inside basket", () => {
  render(
    <Basket items={[sampleProduct]} removeFromBasket={mockRemoveFromBasket} />
  );

  // Ensure the "Remove" button is present
  expect(screen.getByText("Remove")).toBeInTheDocument();

  // Click the "Remove" button
  fireEvent.click(screen.getByText("Remove"));

  // Check if removeFromBasket was called with the correct keyProp
  expect(mockRemoveFromBasket).toHaveBeenCalledWith("sample_key");
});

// Test to check if the correct total amount is displayed for products in the basket
test("displays correct total in the basket", () => {
  render(
    <Basket items={sampleProducts} removeFromBasket={mockRemoveFromBasket} />
  );

  // Calculate the expected total
  const expectedTotal = sampleProducts.reduce(
    (total, item) => total + parseFloat(item.monthlyCost || 0),
    0
  );

  // Ensure each product is displayed with its name, cost, and remove button
  sampleProducts.forEach((item) => {
    expect(
      screen.getByText(`${item.productName} - £${item.monthlyCost} per month`)
    ).toBeInTheDocument();

    const removeButton = screen.getByTestId(`remove-button-${item.keyProp}`);

    // Click the "Remove" button
    fireEvent.click(removeButton);

    // Check if removeFromBasket was called with the correct keyProp
    expect(mockRemoveFromBasket).toHaveBeenCalledWith(item.keyProp);
  });

  // Ensure the total is displayed correctly
  expect(
    screen.getByText(`Total Amount: £${expectedTotal.toFixed(2)}`)
  ).toBeInTheDocument();
});

// Test to make sure the correct product category header is displayed
test("renders ProductCategory with the correct name", () => {
  // Define the category name you want to test
  const categoryName = "Sports";

  // Render the ProductCategory component with the specified name
  render(<ProductCategory name={categoryName} />);

  // Check if the rendered component contains the correct category name
  expect(screen.getByText(categoryName)).toBeInTheDocument();
});

// Test to check if the basket displays correct information for products
test("displays correct content for non-empty basket", () => {
  const sampleItem = {
    keyProp: "sample_key",
    productName: "Sample Product",
    monthlyCost: 19.99,
  };

  render(
    <Basket items={[sampleItem]} removeFromBasket={mockRemoveFromBasket} />
  );

  // Check if the content for a non-empty basket is displayed using data-test-id
  expect(
    screen.getByTestId(`basket-item-${sampleItem.keyProp}`)
  ).toBeInTheDocument();
});

// Test to check if the basket is empty and displays correctly
test("does not display content for empty basket", () => {
  render(<Basket items={[]} removeFromBasket={mockRemoveFromBasket} />);

  // Check if the content for an empty basket is not present using data-test-id
  expect(screen.queryByTestId("empty-basket-message")).not.toBeInTheDocument();
});

// Test the app renders without crashing
test("renders App component without crashing", () => {
  const root = document.createElement("div");
  root.id = "root";
  document.body.appendChild(root);

  render(
    <React.StrictMode>
      <App />
    </React.StrictMode>,
    root
  );

  // Check if specific elements are present in the document
  expect(screen.getByText("Sky")).toBeInTheDocument();
  expect(screen.getByText("Sky F1")).toBeInTheDocument();
  expect(screen.getByText("Sky Golf")).toBeInTheDocument();
  expect(screen.getByText("Sky Tennis")).toBeInTheDocument();
  expect(screen.getByText("Sky Football")).toBeInTheDocument();
  expect(screen.getByText("Sky Horror")).toBeInTheDocument();
  expect(screen.getByText("Sky Cinema")).toBeInTheDocument();
  expect(screen.getByText("Basket")).toBeInTheDocument();
});

// Snapshot test to capture the rendered output component and compare it to stored "snapshot" to ensure that it remains consistent over time.
test("ProductCard snapshot", () => {
  // Render the component
  const { asFragment } = render(
    <ProductCard
      productName="Sample Product"
      description="This is a sample product."
      monthlyCost={19.99}
      contractLength={12}
      src="path/to/image.png"
      type="Sample Type"
      keyProp="sample_key"
      addToBasket={() => {}}
      removeFromBasket={() => {}}
      isInBasket={false}
    />
  );

  // Compare the rendered component with the stored snapshot
  expect(asFragment()).toMatchSnapshot();
});
