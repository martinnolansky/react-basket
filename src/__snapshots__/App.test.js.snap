// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard snapshot 1`] = `
<DocumentFragment>
  <div
    class="max-w-sm bg-white border border-gray-200 rounded-lg shadow dark:bg-gray-800 dark:border-gray-700"
  >
    <img
      alt="Sample Product Logo"
      class="rounded-t-lg mx-auto"
      src="path/to/image.png"
      style="max-width: 100%; height: auto;"
    />
    <div
      class="p-5"
    >
      <h5
        class="mb-2 text-2xl font-bold tracking-tight text-gray-900 dark:text-white"
      >
        Sample Product
      </h5>
      <p
        class="mb-3 font-normal text-gray-700 dark:text-gray-400"
      >
        This is a sample product.
      </p>
      <p
        class="mb-3 font-normal text-gray-700 dark:text-gray-400"
      >
        Monthly Cost: Â£19.99
      </p>
      <p
        class="mb-3 font-normal text-gray-700 dark:text-gray-400"
      >
        Contract Length: 12 months
      </p>
      <button
        class="inline-flex items-center px-3 py-2 text-sm font-medium text-center text-white bg-blue-700 rounded-lg hover:bg-blue-800 focus:ring-4 focus:outline-none focus:ring-blue-300 dark:bg-blue-600 dark:hover:bg-blue-700 dark:focus:ring-blue-800"
      >
        Add to Basket
      </button>
    </div>
  </div>
</DocumentFragment>
`;
